### PR TITLE
[Clojure] Add model generation and conforming

### PR DESCRIPTION
**Pull request by f-f**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

This big PR adds support for generating Clojure Specs from OpenAPI Models.
[Spec](https://clojure.org/guides/spec) is a core library available since Clojure 1.9. Description from the linked page:
> The `spec` library specifies the structure of data, validates or destructures it, and can generate data based on the spec.

Things going on in this PR:
- **Update of the Clojure version `1.7.0` -> `1.9.0`**: This is backwards compatible, so users can just bump their version and everything should be fine.
- **Generate `spec` files from Models**: they are in the `specs/` folder, and imported by all `api` files. This is done with the Data Specs from the [`spec-tools`](https://github.com/metosin/spec-tools#data-specs) library, as they are much nicer to read/write than vanilla specs.
- **Add types to `api` functions via [`defn-spec`](https://github.com/jeaye/orchestra#defn-spec)**: in this way at development time all the inputs and outputs of function calls will be checked and will except if they don't conform (you need to call `orchestra.spec.test/instrument` for it to be active).
- **Add conforming of data returned from the API to the proper model/spec**: this includes e.g. creating date objects instead of getting back strings.
- **Add `:decode-models` setting in `api-context`**: so the above conforming can be turned off if it's indesiderable. This is opt-out (but we can make it opt-in if we want to keep backwards compatibility).

I am not entirely sure this is correct in all cases (e.g. I'm suspicious about multiple arities), but we have an API much bigger than Petstore on which this is already deployed, and the generated code looks fine and works. 

/cc @wing328 @xhh 

Moreover, if this PR goes in I'd be available to be in the Clojure generator "technical committee", as the list is currently empty, and I could help with it.

### CODE REVIEW

**Comment by wing328** [comment]
> Add conforming of data returned from the API to the proper model/spec: this includes e.g. creating date objects instead of getting back strings.

If I understand this correctly, this is a breaking change and therefore I've labeled this PR as such.

I'm testing the change and will merge when the test passes.

File: samples/client/petstore/clojure/pom.xml [hunk 1]
```diff
@@ -1,32 +0,0 @@
-<project>
```
Comment (wing328) [inline samples/client/petstore/clojure/pom.xml]: @f-f please add back pom.xml so that CI (circleci) can run the tests for Clojure Petstore sample

Comment (f-f) [reply samples/client/petstore/clojure/pom.xml]: Yes sorry, I accidentally deleted them.

Comment (wing328) [reply samples/client/petstore/clojure/pom.xml]: No problem. Let us know if you need help running the tests locally for verification.

**Comment by f-f** [comment]
> If I understand correctly, this is a breaking change and therefore I've labeled this PR as such

It is if the `:decode-models` setting is `true` (which is the current state). If not, the code behaves as it was before. If we want to keep backwards compatibility we can default it to `false`, and make this new behaviour opt-in.

**Comment by wing328** [comment]
Thanks for the explanation. I would suggest defaulting `:decode-models` to `false` to make it backward compatible.

**Comment by f-f** [comment]
👍 I pushed the change to the setting and the updated sample.

However, the tests are failing because we are on Java7, but this PR requires Java8. How can I upgrade it?

**Comment by jmini** [comment]
The generator itself now requires java 8 so I think that it is fine to upgrade it.

**Comment by wing328** [comment]
>However, the tests are failing because we are on Java7, but this PR requires Java8. How can I upgrade it?

Please remove https://github.com/f-f/openapi-generator/blob/2750e249a08956ad67d87585d832b94813555b28/CI/pom.xml.circleci.java7#L839

**Comment by f-f** [comment]
@wing328 I pushed a fix for the tests, but they don't pass on my machine due to the API returning a 500 (which happens when I run the tests on master). They also don't pass on the CI seemingly due to the same issue (but CI log doesn't show the exceptions?).

I'm testing with running `mvn verify` inside the `samples/client/petstore/clojure`. Am I missing any step?

**Comment by wing328** [comment]
@f-f the CircleCI tests failed with the following message:
```
lein test open-api-petstore.api.pet-test

lein test open-api-petstore.api.store-test

lein test :only open-api-petstore.api.store-test/test-place-and-delete-order

FAIL in (test-place-and-delete-order) (store_test.clj:31)
expected: (= (attr order) (attr fetched))
  actual: (not (= 200 nil))

lein test open-api-petstore.api.user-test

lein test :only open-api-petstore.api.user-test/test-create-and-delete-user

FAIL in (test-create-and-delete-user) (user_test.clj:27)
expected: (= (attr user) (attr fetched))
  actual: (not (= 0 nil))

lein test open-api-petstore.core-test
```

I'll mark this as a breaking change without fallback as the Clojure client no longer works with JDK7

**Comment by wing328** [comment]
One way is to run the petstore server locally: https://hub.docker.com/r/swaggerapi/petstore/

and you will need to update the host table:
```
127.0.0.1 petstore.swagger.io
```

**Comment by f-f** [comment]
@wing328 thanks a lot!

I fixed the existing tests, and added a test for the new decoding feature; now all tests pass locally.

**Comment by f-f** [comment]
@wing328 yeah I saw that and it's failing because of not finding the classes, which the last commit I pushed (88726c1655b0294fa44e14334fe7983e28982684) should have fixed (needs the capitalized filenames).
However, on my mac I have a case-insentitive HFS, so I'm not able to push the updated petstore sample with the capitalized filenames (but the tests pass on my machine). I'll fix as soon I get to my linux machine.

**Comment by f-f** [comment]
I found a small bug in the meanwhile, so I'll ping again here once it's fixed

**Comment by wing328** [comment]
> I found a small bug in the meanwhile, so I'll ping again here once it's fixed

Thanks for the update. If you need help from us, please let us know.

Since it's a breaking change, I've changed the target branch to be "4.1.x" (for breaking changes without fallback)

**Comment by langford** [comment]
@f-f Thank you for taking the time to do this work, and share it back to the community. 

I'm seeing this working without test failures on MacOSX, does this fail anywhere other than CI reliably?

**Comment by langford** [comment]
Also, @f-f, you should be able to use 

        git config core.ignorecase false

in the project directory to fix your issue with case insensitivity 

https://stackoverflow.com/a/17688308

**Comment by wing328** [comment]
@f-f you can consider using [fake pestore spec](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml) instead of petstore.yaml/json as fake petstore covers a lot more edge cases (which may not be all addressed by this PR)

**Comment by langford** [comment]
>This is the current state, and works nicely for the Petstore. However it stopped working for our internal project, because we have a model called Package. Turns out, in Clojure you cannot define a Package var, because the name is taken:

Would prefixing all the spec names with a long hardcoded string allow this to be verified as working and merged sooner? It seems like that step could allow this all to be validated without worrying about removing the capitalization code. 

I offer this alternative as it seems this would also be less mysterious than context-dependent renames for people using the plugin on their own codebases. I doubt many people would go "of course my model name collides with the java keyword "strictfp", that's why it renames. 

Example:

        oags-spec-Petstore, oags-spec-Package

> more complex example API to test the program on

Checking large specs is a great idea. Other specs that might work for validation purposes: 

YAML, possibly handwritten: 
https://github.com/stripe/openapi/blob/master/openapi/spec3.yaml

JSON generated: 
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/billing/resource-manager/Microsoft.Billing/preview/2018-03-01-preview/billing.json

Several different APIs using many features (2.0, but seems like they should be compatible):
https://gist.github.com/arno-di-loreto/5a3df2250721fb154060#file-simple_openapi_specification_22_documentation_with_gfm-yaml

**Comment by wing328** [comment]
@f-f is there anything we (the community) can help with this PR?

**Comment by f-f** [comment]
I had the fix on our patched version since a while, but wanted to test it and think about it a bit more.  
Anyways I just pushed it here, and it passes the tests on the petstore and on our internal (quite complex) API, so it's good to go by me :)

Things I did to fix:
- uncapitalize the generated specs, so they are not mixed up with classnames
- added a postfix to the spec names, so they don't risk being mixed up with other names, as @langford suggested

@wing328 
> is there anything we (the community) can help with this PR?

Yeah, as I mentioned before we should setup more complex tests. I opened #743 to track that

**Comment by wing328** [comment]
@f-f I'll merge it into 4.0.x tomorrow (Sunday) if no one has further feedback on this PR.

**Comment by wing328** [comment]
@f-f PR merged into 4.0.x branch. Thanks for your contribution 👍

**Comment by f-f** [comment]
@wing328 thank you! :)

**Pull request merged**